Fix hypernoblium formation gas reaction rate sanity clamping not accounting for the reduction factor. Also other sanity fixes.

Pickle-Coding opened the pull request:
## About The Pull Request
Fix hypernoblium formation not accounting for the reduction factor when sanity clamping the reaction rate.

Fix N2O formation not using updated consumption rates for sanity checks.

Fix BZ formation not accounting for nitrous oxide decomposition factor for reaction rate sanity clamping.
## Why It's Good For The Game
So the reactions are more natural.
## Changelog
:cl:
fix: Fix hypernoblium formation gas reaction rate sanity clamping not accounting for the reduction factor.
fix: Fix N2O formation gas reaction rate clamping clamping for incorrect consumption rates.
fix: Fix BZ formation reaction rate sanity clamping not accounting for N2O decomposition factor.
/:cl:

Pickle-Coding committed:
Fix hypernoblium formation reaction rate sanity.

Fixes hypernoblium formation reaction rate mole count sanity min check. Does this by calculating the bz reduction factor before reaction rate (fuck, I am on mobile and I can not see what I am typing, also I kinda forgot what I said AAAAAAAAAAAA) calculating the reaction rate. It then applies the product of the tritium consumption rate and reduction factor thinf uhh it sucks typing without seeing what I'm writing AAAAAAAA thrhfhfjjgjjhdhshgfsjsbdhhdbzjabdjahsjcnakdbd

Pickle-Coding committed:
Fix N2O formation sanity checks.

Fixes N2O formation sanity checks by making them correctly scale with their consumption rate. This was probably my fault like a year ago when I changed shit.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -448,8 +448,8 @@
 	var/list/cached_gases = air.gases
-	var/heat_efficency = min(cached_gases[/datum/gas/oxygen][MOLES], cached_gases[/datum/gas/nitrogen][MOLES] * INVERSE(2))
-	if ((cached_gases[/datum/gas/oxygen][MOLES] - heat_efficency < 0 ) || (cached_gases[/datum/gas/nitrogen][MOLES] - heat_efficency * 2 < 0))
+	var/heat_efficency = min(cached_gases[/datum/gas/oxygen][MOLES] * INVERSE(0.5), cached_gases[/datum/gas/nitrogen][MOLES])
+	if ((cached_gases[/datum/gas/oxygen][MOLES] - heat_efficency * 0.5 < 0 ) || (cached_gases[/datum/gas/nitrogen][MOLES] - heat_efficency < 0))
 		return NO_REACTION // Shouldn't produce gas from nothing.
```
undocced?

Pickle-Coding replied:
Mobile copy paste failure moment.

Pickle-Coding replied:
Wait nevermind. What do you mean by undocced? What got undocced?

LemonInTheDark replied:
you hadn't mentioned this change in the desc when I got to it, so I got confused is all

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -779,14 +779,17 @@
-	var/nob_formed = min((cached_gases[/datum/gas/nitrogen][MOLES] + cached_gases[/datum/gas/tritium][MOLES]) * 0.01, cached_gases[/datum/gas/tritium][MOLES] * INVERSE(5), cached_gases[/datum/gas/nitrogen][MOLES] * INVERSE(10))
-
-	if (nob_formed <= 0 || (cached_gases[/datum/gas/tritium][MOLES] - 5 * nob_formed < 0) || (cached_gases[/datum/gas/nitrogen][MOLES] - 10 * nob_formed < 0))
+	/// List of gases we will assert, and possibly garbage collect.
+	var/list/asserted_gases = list(/datum/gas/hypernoblium, /datum/gas/bz)
+	air.assert_gases(arglist(asserted_gases))
```
annoying

Pickle-Coding replied:
Guh?

Pickle-Coding committed:
BZ formation reaction rate sanity clamping fix.

Fixes bz formation reaction rate sanity clamping to account for nitrous oxide decomposed factor.

Pickle-Coding committed:
Forgot.

Apply bitrous oxide decomposed factor to the stupid early return sanity if check.

EliteCreature commented:
When you say 'Sanity Clamping', what does that mean? [Just my first time hearing this phrase]

Pickle-Coding commented:
Clamp reaction rate to not consume more moles than there is, so no negative moles or creating gas from nothing bullshit.
